### PR TITLE
[C] Port the specificity concept to Bindings

### LINKED ISSUE #16538: Setting button IsEnabled to 'true' does not dynamically change background color to match dark/light theme
### Description

I have an issue:

- When an application has two buttons where one is enabled and one is disabled
- When the disabled button gets programmatically enabled, it looks fine.
- However, once the user dynamically changes the theme to light or dark, the button theme doesn't change to the new dark/light mode.

Real life scenario:

- I have an application that requires a user to sign in, in order to use certain features. By default, some buttons are disabled until a successful authentication completes. By default, the application is in light mode.
- Once authentication completes, those buttons are enabled. It looks fine.
- On the top of my application, I have button that will toggle between light and dark mode. Once the user clicks on the button to toggle into dark mode, the newly enabled buttons are still left in light mode style.

https://github.com/dotnet/maui/assets/69189298/6d73d187-1f8c-4a2e-a512-4224171b84a5

### Steps to Reproduce

1. Create new MAUI application.
2. Open MainPage.xaml that was auto generated. We will modify everything between the <ContentPage></ContentPage> to this:

```
    <ContentPage.Resources>
        <x:String x:Key="SwitchToLightString">Switch to Light Mode</x:String>
        <x:String x:Key="SwitchToDarkString">Switch to Dark Mode</x:String>

    </ContentPage.Resources>

    <VerticalStackLayout>
        <Button x:Name="btnSwitchMode"
                Clicked="btnSwitchMode_Clicked"
                Margin="0,2,0,2"
                Text="{AppThemeBinding Dark={StaticResource SwitchToLightString}, Light={StaticResource SwitchToDarkString}}"/>

        <Button x:Name="btnEnableButtonBelow"
                Clicked="btnEnableButtonBelow_Clicked"
                Margin="0,2,0,2"
                Text="Enable Button Below"/>

        <Button x:Name="btnInitiallyDisabledButton"
                IsEnabled="False"
                Margin="0,2,0,2"
                Text="Initially Disabled Button"/>

    </VerticalStackLayout>
```

3. Then, we will change everything in the MainPage.xaml.cs file to this:
```
public partial class MainPage : ContentPage
{
	public MainPage()
	{
		InitializeComponent();

        App.Current.UserAppTheme = AppTheme.Dark;
	}

    void btnSwitchMode_Clicked(System.Object sender, System.EventArgs e)
    {
        if (App.Current.UserAppTheme is AppTheme.Dark)
            App.Current.UserAppTheme = AppTheme.Light;
        else
            App.Current.UserAppTheme = AppTheme.Dark;
    }

    void btnEnableButtonBelow_Clicked(System.Object sender, System.EventArgs e)
    {
        btnInitiallyDisabledButton.IsEnabled = true;
    }
}
```

4. Run the application in Debug mode. It will be set at a dark mode (we set that in the constructor)
5. Click on the middle button to enable to last button. Everything should still look fine.
6. Click on the top button to change to light mode. Now you will see the issue - the second/middle button and the last button are both enabled. However, they have different styles.

### Link to public reproduction project repository

_No response_

### Version with bug

7.0.92

### Last version that worked well

7.0.92

### Affected platforms

macOS

### Affected platform versions

MacOS Ventura 13.4.1

### Did you find any workaround?

Do not allow the user to dynamically change the theme. You can allow the user to change the theme, but require the application to be closed and reopened in order for the application theme to update to the requested dark/light theme.

### Relevant log output

_No response_

### LINKED ISSUE #17354: [regression/8.0.0-preview.7.8842] VisualState does not return to `Normal` state when using an `AppThemeBinding`
### Description

Utilizing a `VisualState` that defines an `AppThemeBinding` in `PointerOver` never returns to `Normal` state. 

### Steps to Reproduce

Given 
```xml
            <Grid WidthRequest="500" ColumnDefinitions="*">
                <VisualStateManager.VisualStateGroups>
                    <VisualStateGroupList>
                        <VisualStateGroup x:Name="CommonStates">
                            <VisualState x:Name="PointerOver">
                                <VisualState.Setters>
                                    <Setter Property="BackgroundColor" Value="{AppThemeBinding Dark=Red, Light=White}"/>
                                </VisualState.Setters>
                            </VisualState>
                            <VisualState x:Name="Normal">
                                <VisualState.Setters>
                                    <Setter Property="BackgroundColor" Value="Transparent" />
                                </VisualState.Setters>
                            </VisualState>
                        </VisualStateGroup>
                    </VisualStateGroupList>
                </VisualStateManager.VisualStateGroups>
                <Label Text="I'm a grid. Hover over me to change background color"/>
            </Grid>
 ```
 
 Observe the following.
 
 1. Hover over the grid. Observe the color change to RED.
 2. No longer hover over the grid. The color still stays RED.
 
 This does **not** occur if you remove the `AppThemeBinding` binding and just use `RED`.

### Link to public reproduction project repository

https://github.com/daltzctr/maui-hoverstate-repro

### Version with bug

[8.0.0-preview.7.8842](https://github.com/dotnet/maui/releases/tag/8.0.0-preview.7.8842)

### Is this a regression from previous behavior?

Yes, this used to work in .NET MAUI

### Last version that worked well

[8.0.0-preview.6.8686](https://github.com/dotnet/maui/releases/tag/8.0.0-preview.6.8686)

### Affected platforms

Windows

### Affected platform versions

_No response_

### Did you find any workaround?

Don't use `AppThemeBinding`. Unfortunately, for applications that support more than one theme (dark and white), this is not an acceptable solution. 

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by StephaneDelcroix**
### Description of Change

Allow having multiple layers of Bindings, useful for e.g. AppThemeBinding, Style, VSM

### Issues Fixed

- fixes #16538
- fixes #17354

### CODE REVIEW

File: src/Controls/src/Core/BindableObject.cs [hunk 1]
```diff
@@ -253,8 +253,19 @@ public void RemoveBinding(BindableProperty property)
...
-			if (context?.Binding != null)
-				RemoveBinding(property, context);
+			var specificity = SetterSpecificity.FromBinding;
+			if (context != null && context.Bindings.Count > 0)
+				specificity = context.Bindings.Last().Key;
+
+			RemoveBinding(property, specificity);
+		}
+
+		internal void RemoveBinding(BindableProperty property, SetterSpecificity specificity)
```
Comment (rmarinho) [inline src/Controls/src/Core/BindableObject.cs:263]: Should this public for Library authors to use ?

File: src/Controls/src/Core/BindableObject.cs [hunk 2]
```diff
@@ -286,11 +297,24 @@ internal void SetBinding(BindableProperty targetProperty, BindingBase binding, S
 				context.Values.Add(SetterSpecificity.FromBinding, kvp.Value);
 			}
 
-			context.Binding?.Unapply();
-			context.BindingSpecificity = specificity;
+			BindingBase oldBinding = null;
+			SetterSpecificity oldSpecificity = default;
+			if (context.Bindings.Count > 0)
```
Comment (rmarinho) [inline src/Controls/src/Core/BindableObject.cs:302]: All of this code is only run on the UIThread right? so we shouldn't have this be called from 2 different threads and one could get the Bindings collection with different values.

File: src/Controls/src/Core/BindableObject.cs [hunk 3]
```diff
@@ -253,8 +253,19 @@ public void RemoveBinding(BindableProperty property)
 		{
 			BindablePropertyContext context = GetContext(property ?? throw new ArgumentNullException(nameof(property)));
 
-			if (context?.Binding != null)
-				RemoveBinding(property, context);
+			var specificity = SetterSpecificity.FromBinding;
+			if (context != null && context.Bindings.Count > 0)
+				specificity = context.Bindings.Last().Key;
```
[the remark this answers is not in the document]
Comment (StephaneDelcroix) [reply src/Controls/src/Core/BindableObject.cs:258]: no, it's a sorted list. last one always win, that's the rule

File: src/Controls/src/Core/BindableObject.cs (hunk 1 above)
Comment (StephaneDelcroix) [reply src/Controls/src/Core/BindableObject.cs:263]: no, specificity isn't public

File: src/Controls/src/Core/BindableObject.cs (hunk 2 above)
Comment (StephaneDelcroix) [reply src/Controls/src/Core/BindableObject.cs:302]: unlikely

**Review by PureWeen — APPROVED** [review]

**Pull request merged**